golang/channel.go

```diff
@@ -231,34 +231,19 @@ func (ch *Channel) createCommonStats() {
 	// TODO(prashant): Allow user to pass extra tags (such as cluster, version).
 }
 
-func (ch *Channel) registerNewSubChannel(serviceName string) *SubChannel {
-	mutable := &ch.mutable
-	mutable.mut.Lock()
-	defer mutable.mut.Unlock()
-
-	// Recheck for the subchannel under the write lock.
-	if sc, ok := mutable.subChannels[serviceName]; ok {
-		return sc
-	}
-
-	sc := newSubChannel(serviceName, ch.peers)
-	mutable.subChannels[serviceName] = sc
-	return sc
-}
-
 // GetSubChannel returns a SubChannel for the given service name. If the subchannel does not
 // exist, it is created.
 func (ch *Channel) GetSubChannel(serviceName string) *SubChannel {
-	mutable := &ch.mutable
-	mutable.mut.RLock()
+	subChMap := ch.subChannels
```

Comment:
This logic should live inside of subChannelMap. I'd probably just make a "getOrAdd" style function this uses.